src/Text/Pretty/Core.agda

```diff
@@ -0,0 +1,343 @@
+
+-- Content is a representation of the first line and the middle of the block.
+-- We use a tree rather than a list for the middle of the block so that we can
+-- extend it with lines on the left and on the line for free. We will ultimately
+-- render the block by traversing the tree left to right in a depth-first manner.
+
+Content : Set
+Content = Maybe (String × Tree String)
+
+size : Content → ℕ
+size = maybe′ (suc ∘ Tree.size ∘ proj₂) 0
+
+All : ∀ {p} (P : String → Set p) → (Content → Set p)
+All P = Allᴹ.All (Allᴾ.All P (Allᵀ.All P))
+
+All≤ : ℕ → Content → Set
+All≤ n = All (λ s → length s ≤ n)
+
+record Block : Set where
+  field
+    height    : ℕ
+    block     : [ xs ∈ Content ∣ size xs ≡ height ]
+  -- last line
+    lastWidth : ℕ
+    last      : [ s ∈ String ∣ length s ≡ lastWidth ]
+  -- max of all the widths
+    maxWidth  : [ n ∈ ℕ ∣ lastWidth ≤ n × All≤ n (block .value) ]
+
+------------------------------------------------------------------------
+-- Raw string
+
+text : String → Block
+text s = record
+  { height    = 0
+  ; block     = nothing , ⦇ refl ⦈
+  ; lastWidth = width
+  ; last      = s , ⦇ refl ⦈
+  ; maxWidth  = width , ⦇ (≤-refl , nothing) ⦈
+  } where width = length s; open Erased
+
+------------------------------------------------------------------------
+-- Empty
+
+empty : Block
+empty = text ""
+
+------------------------------------------------------------------------
+-- Helper functions
+
+node? : Content → String → Tree String → Content
+node? (just (x , xs)) y ys = just (x , node xs y ys)
+node? nothing         y ys = just (y , ys)
+
+∣node?∣ : ∀ b y ys → size (node? b y ys)
+                   ≡ size b + suc (Tree.size ys)
+∣node?∣ (just (x , xs)) y ys = refl
+∣node?∣ nothing         y ys = refl
+
+≤-Content : ∀ {m n} {b : Content} → m ≤ n → All≤ m b → All≤ n b
+≤-Content {m} {n} m≤n = Allᴹ.map (Prod.map step (Allᵀ.map step)) where
+
+  step : ∀ {p} → p ≤ m → p ≤ n
+  step = flip ≤-trans m≤n
+
+All≤-node? : ∀ {l m r n} →
+  All≤ n l → length m ≤ n → Allᵀ.All (λ s → length s ≤ n) r →
+  All≤ n (node? l m r)
+All≤-node? nothing           py pys = just (py , pys)
+All≤-node? (just (px , pxs)) py pys = just (px , node pxs py pys)
+
+------------------------------------------------------------------------
+-- Appending two documents
+
+private
+  module append (x y : Block) where
+
+    module x = Block x
+    module y = Block y
+
+    blockx = x.block .value
+    blocky = y.block .value
+
+    widthx = x.maxWidth .value
+    widthy = y.maxWidth .value
+
+    lastx = x.last .value
+    lasty = y.last .value
+
+    height : ℕ
+    height = (_+_ on Block.height) x y
+
+    lastWidth : ℕ
+    lastWidth = (_+_ on Block.lastWidth) x y
+
+    pad : Maybe String
+    pad with x.lastWidth
+    ... | 0 = nothing
+    ... | l = just (replicate l ' ')
+
+    size-pad : maybe′ length 0 pad ≡ x.lastWidth
+    size-pad with x.lastWidth
+    ... | 0         = refl
+    ... | l@(suc _) = length-replicate l
```

Comment:
Presumably these will go once #994 gets merged in?

Reply:
I would have to look into it. Given that we use the same pad for multiple
lines, I would really like to generate the string only once and use it
for all of them.